### DISCUSSION TITLE
Automatic hyperlinks for urls in log event context properties

### DISCUSSION THREAD

**Discussion by laurencee** [datalust/seq-tickets#1115] [Ideas]
Hello,

After using Seq for a while at various places I feel like it'd be a nice improvement if urls were automatically converted to hyperlinks in context properties. I have found it to be quite common that I make use of context properties to link to external places related to a log event.

I'm not sure of the negatives of doing something like this but it would improve my user experience by not having to select urls out of the text fields and open them in a new tab (making sure I selected all the text for the link and what not).

Maybe this could be extended to the messages as well? That could cause some issues with attempting to expand a log event however which is why I'm being specific to context properties.

Let me know your thoughts!

**Comment by nblumhardt** [comment]
Thanks for the note, Laurence! We've considered this in the past - we're just a little uneasy about the implications for security; even if we mark the links `rel="noopener noreferrer"`, it still seems like having clickable URLs that could come from untrusted sources in the UI is questionable. (E.g. a malicious user might enter the URL as their Name in a web form, which would result in a hyperlinked value in the Name property if logged.)

Moving to a discussion, since we're now tracking uncommitted features over there :+1:

**Reply by jamesfera** [reply]
I see a lot of value to this feature.  This would allow deep linking directly into problem areas surfaced by the logs.

Perhaps allowing admins to provide a whitelist mask (eg: *.domain.com) would mitigate some security issues.

**Comment by liammclennan** [comment]
Links back to Seq are also an issue for CSRF.

**Reply by jdmorseyyc** [reply]
We could use this feature to link from certain properties (correlation ID for example) to our other services.

**Comment by ergunozyurt** [comment]
This kind of action is quite possible within [Elastic APM](https://www.elastic.co/guide/en/kibana/current/custom-links.html) and [Grafana](https://grafana.com/docs/grafana/latest/linking/data-links/) and they help us big time on navigating between those tools while staying in the same context / filters.

We link back to our Seq instance from ElasticAPM and Grafana to connect traces and dashboards and signals at the moment. Would be awesome if we could also do the other way around.

**Reply by nblumhardt** [reply]
Thanks for chiming in! The features of those products are slightly different; they allow links to be constructed from event properties - not hyperlinking of arbitrary URLs found in event properties.

A Seq version of the feature might look like defining some templates for different URLs where certain properties exist. E.g. if you include an `Application` property in your events, you might define a link template within Seq like:

| Link title | URL template |
| --- | --- |
| Attach to new GitHub issue | `https://github.com/datalust/{Application}/issues/new?title={@Message}` |

This would then appear in e.g. an "Actions" or "Links" drop down on expanded events:

![image](https://user-images.githubusercontent.com/342712/165864480-b00dd924-8543-45b0-b787-06e8a3891bc3.png)

While pretty tricky to implement, I think this design would avoid the worst security concerns. Would love to know everyone's thoughts!

**Reply by ergunozyurt** [reply]
> Thanks for chiming in! The features of those products are slightly different; they allow links to be constructed from event properties - not hyperlinking of arbitrary URLs found in event properties.

That's correct and it's more inclusive in that sense; an arbitrary URL in the url can directly be targeted without an addition but this isn't an issue until the template has been set that way intentionally. 
 
Good thing with this approach is that the responsibility is on the viewer admin side so the log producer has a limited capability on the end result.

I like the suggested implementation!
